crypto/evp/p_lib.c

```diff
@@ -665,6 +665,25 @@ int EVP_PKEY_get_default_digest_nid(EVP_PKEY *pkey, int *pnid)
     return evp_pkey_asn1_ctrl(pkey, ASN1_PKEY_CTRL_DEFAULT_MD_NID, 0, pnid);
 }
 
+int EVP_PKEY_supports_digest_nid(EVP_PKEY *pkey, int nid)
+{
+    int rv, default_nid;
+
+    rv = evp_pkey_asn1_ctrl(pkey, ASN1_PKEY_CTRL_SUPPORTS_MD_NID, nid, NULL);
+    if (rv < 0)
```

Comment:
Should this explicitly check for -2 rather than <0 and threat everything else as an error?

Also, I believe we should add developer documentation about the expectations for an implementation of a pkey_ctrl supporting ASN1_PKEY_CTRL_SUPPORTS_MD_NID

Reply:
> Should this explicitly check for `-2` rather than `<0` and treat everything else as an error?

Ultimately we have to decide whether to use the given hash, or not. There *isn't* a common failure mode where it returns anything but -2, is there? Either it gets all the way through to the `ameth` which can return an appropriate value in the [0,1] range, or it doesn't. And if it doesn't then that is basically the same as the `-2` error: we don't know the answer, and we should probably try it anyway (because we sure as hell don't want to refuse to try *anything*).

> Also, I believe we should add developer documentation about the expectations for an implementation of a `pkey_ctrl` supporting `ASN1_PKEY_CTRL_SUPPORTS_MD_NID`.

Ack. I'll also split it out into separate commits to fix the EC handling of `DEFAULT_MD_NID`, then use `DEFAULT_MD_NID` in the TLS code, then finally add `SUPPORTS_MD_NID`. The first two can and should be backported to older releases as bug fixes.

Reply:
> Ultimately we have to decide whether to use the given hash, or not. There _isn't_ a common failure mode where it returns anything but -2, is there? Either it gets all the way through to the `ameth` which can return an appropriate value in the [0,1] range, or it doesn't. And if it doesn't then that is basically the same as the `-2` error: we don't know the answer, and we should probably try it anyway (because we sure as hell don't want to refuse to try _anything_).

As you said, calling [`evp_pkey_asn1_ctrl`](https://github.com/openssl/openssl/blob/32f803d88ec3df7f95dfbf840c271f7438ce3357/crypto/evp/p_lib.c#L656) will end up either answering `-2` (_"not implemented"_) or calling through the `ameth` with the `ASN1_PKEY_CTRL_SUPPORTS_MD_NID` command, which should itself return `[0, 1, -2]` ([_"false"_, _"true"_, _"not supported"_]).
If the `ameth` for `ASN1_PKEY_CTRL_SUPPORTS_MD_NID` returns any other value, it's either a programmer error in implementing the custom `ameth` or the `ameth` needed to signal an error: in either case I don't think we should silently ignore it.

Reply:
Normally I'm the first to rant about poor error handling, but on this occasion I'm just not sure it's worth it. If there's a genuine problem with the signature, that'll fail soon enough when we try it. There's nothing useful that a caller can do with an error here, apart from try it anyway... which is what I've implemented.

Reply:
I disagree: if an error (`!= - 2`) is returned for any reason, and we don't report it as soon as it happens, if later there is a failure, it might be quite hard to debug where the first failure happened.